### PR TITLE
Adds Bluespace Inhibitors (Telescience Nerf)

### PR DESCRIPTION
**Pull request by BurgerLUA**
# Overview

Adds telescience inhibitors to the station. These are fragile, powered machines that inhibit where bluespace activity can occur. The way it works is that it creates an 8 radius (16 tile wide) circular field that displaces any incoming or outgoing bluespace signal to an unintended destination just outside the area of effect. 

Basically, incoming teleport signals are redirected to around a random location around the circle, and outgoing teleport signals are redirected to around a random location around the circle. For example, if a ninja was inside the area of effect and tried to teleport outside of it, it wouldn't work but instead teleport something else around me just outside the circle. This will very rarely occur, so most that will happen is that the ninja won't be able to teleport while inside the area of effect. If a ninja was outside the area of effect and tried teleporting inside the circle, they would instead teleport to a random location outside the circle.

A semi-intended side effect is that it would be harder for ninjas/wizards to break into the armory/vault/ai core. Of course, they could use other tools and walk inside the vault. This does not stop them from ever entering it, but makes it unsafe to teleport out.

EMPing a bluespace inhibitor teleports everything in a 20 tile radius to a random location on the station. This teleportation is "safe" and will only teleport the objects in a random station area in unoccupied space.

Feedback Thread: https://forums.aurorastation.org/viewtopic.php?f=18&t=12285

### IC changelog
`NanoTrasen's Telescience Sector has recently unveiled a new bluespace prototype designed to inhibit and manipulate incoming or outgoing bluespace signals, no doubt as a response to the recent strings of high-value tele thefts that have occured by stolen teleportation technology.

An private investor conference was held earlier today to explain the change in practical terms, as well as future prospects for bluespace technology. Philanthropist and investor Richard Hollande paraphrased what the investor had to say in a comment to the press.

"Think of it like this. I have a machine that generates some sort of bubbleshield forcefield. Someone throws a rock, the rock being the incoming bluespace signal, and that rock is thrown at the bubble shield." Richard Hollande continued, "The rock will bounce off and land somewhere else; somewhere uncalculated by quantum mechanics, they said "That rock analogy works both ways. If someone tries to grab something through [bluespace], their hand will bounce off and grab something else."

It is not without its critics, however. Leaders of fringe xenophobic conspiracy groups claim that they're monopolizing on problems caused by the Skrell. Unathi Military investors claim that the design is too fragile for use in ramming speed tactics used by their flagships.

NanoTrasen has not released a public statement about the design, yet.`

### CODE REVIEW

**Comment by BurgerLUA** [comment]
!review

**Comment by BurgerLUA** [comment]
@skull132 (It actually does.)

Anything that uses the do_teleport framework is supported by this.

**Comment by BurgerLUA** [comment]
It will only break during an explosion however I think that would not be enough.
I will be making it so that it instantly breaks if any damage is taken.

File: code/datums/helper_datums/teleport.dm [hunk 1]
```diff
@@ -39,22 +39,50 @@
...
+		for(var/found_inhibitor in circlerange(adestination,8))
+			if(!istype(found_inhibitor,/obj/machinery/anti_bluespace))
+				continue
+			var/obj/machinery/anti_bluespace/AB = found_inhibitor
+			if(AB.stat & (NOPOWER | BROKEN) )
+				continue
+			bad_turfs += circlerangeturfs(get_turf(AB),8)
+			good_turfs += circlerangeturfs(get_turf(AB),9)
+		if(good_turfs.len && bad_turfs.len)
+			good_turfs -= bad_turfs
```
Comment (Sindorman) [inline code/datums/helper_datums/teleport.dm]: @skull132 tell me if this is expensive, I cannot tell. good_turfs is of 251 tiles and bad_turfs is 201 tiles. Is it worth writing proc that gets just the perimeter tiles? Or making it work on map AREA?

File: code/game/machinery/anti_bluespace.dm [hunk 2]
```diff
@@ -0,0 +1,97 @@
...
+		return
+	playsound(src.loc, 'sound/effects/grillehit.ogg', 100, 1)
+	visible_message(span("warning","\The [src] breaks!"))
+	stat |= BROKEN
+	anchored = 0
+	update_icon()
+
+/obj/machinery/anti_bluespace/attackby(obj/item/weapon/W as obj, mob/user as mob)
+	if(user.a_intent == I_HURT)
+		visible_message(span("notice","\The [user] touches \the [src] with \the [W]."))
```
Comment (Sindorman) [inline code/game/machinery/anti_bluespace.dm]: This should be swapped with line 55

File: code/datums/helper_datums/teleport.dm (hunk 1 above)
Comment (Sindorman) [reply code/datums/helper_datums/teleport.dm]: I Guess it should be fine since it is not called often

**Comment by BurgerLUA** [comment]
Changes:
Bullets that do damage will only break it.
You can break it by attacking it with your hand, as long as you are a carbon based lifeform. 
Using items or your hand with help intent on has a 66% chance to not break it.

File: code/game/machinery/anti_bluespace.dm [hunk 3]
```diff
@@ -0,0 +1,115 @@
+/obj/machinery/anti_bluespace
+	name = "bluespace inhibitor"
+	desc = "Scrambles any bluespace related activity and displaces it away from the beacon's area of effect."
+	icon = 'icons/obj/telescience.dmi'
+	icon_state = "nopad"
+	anchored = 1
+	density = 1
+	use_power = 1
+	idle_power_usage = 500
```
Comment (Sindorman) [inline code/game/machinery/anti_bluespace.dm]: I would make it consume way more power

Comment (BurgerLUA) [reply code/game/machinery/anti_bluespace.dm]: I think the power level is fine however if there is a balancing issue with it, I will consider changes to power consumption.

File: code/modules/spells/general/mark_recall.dm [hunk 4]
```diff
@@ -36,7 +36,7 @@
 	var/turf/T = get_turf(target)
 	if(!T)
 		return 0
-	user.forceMove(T)
+	do_teleport(user,T)
```
Comment (Alberyk) [inline code/modules/spells/general/mark_recall.dm]: Nope, wizard powers should not be forced to suffer the same restrictions as telescience.

Comment (BurgerLUA) [reply code/modules/spells/general/mark_recall.dm]: The wizard should not be able to teleport into the vault tbh.

Comment (Alberyk) [reply code/modules/spells/general/mark_recall.dm]: It is magic, not bluespace.

File: code/modules/spells/general/mark_recall.dm [hunk 5]
```diff
@@ -36,7 +36,7 @@
 	var/turf/T = get_turf(target)
 	if(!T)
 		return 0
-	do_teleport(user,T)
+	user.forceMove(T)
```
Comment (LordFowl) [inline code/modules/spells/general/mark_recall.dm]: The wizard should be held to the same restrictions as other teleporting antagonists. Why is the wizard the only antag that can just warp into the vault/the armory/the bridge for free loot at round-start?

Comment (Alberyk) [reply code/modules/spells/general/mark_recall.dm]: That is not how this works, you need to reach the place where you want to set the mark, so, if you want to teleport back to the vault using this, you had to manage breaking into it first.

Comment (BurgerLUA) [reply code/modules/spells/general/mark_recall.dm]: PLEASE COME UP WITH A SOLUTION BETWEEN YOURSELVES

**Pull request merged**